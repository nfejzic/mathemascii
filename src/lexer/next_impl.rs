macro_rules! next_impl {
    ($self: ident, no_prefix: $($infallible_fn:ident),*; prefix: $($fallible_fn:ident),* or $fallback_fn:ident) => {{
        let mut curr = $self.curr;
        let mut token = None;

        $(
        if let Some((num, cursor)) = $self.$infallible_fn() {
            // number is never prefix of anything
            $self.curr = cursor;
            return Some(num);
        }
        )*

        $(
        if let Some((found_token, cursor)) = $self.$fallible_fn(curr - $self.curr + 1) {
            token = Some(found_token);
            curr = cursor;

            // token can't contain a whitespace, so this token has the maximum length already
            if $self.src.get(cursor).map(|s| s.is_whitespace()).unwrap_or(false) {
                $self.curr = curr;
                return token;
            }
        }
        )*

        match token {
            Some(token) => {
                $self.curr = curr;
                Some(token)
            }
            None => {
                let (token, curr) = $self.$fallback_fn(curr - $self.curr + 1)?;
                $self.curr = curr;
                return Some(token);
            }
        }
    }};
}

pub(super) use next_impl;
