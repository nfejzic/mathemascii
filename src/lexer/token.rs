use super::{
    keywords::{
        accents::Accent, arrows::Arrow, font_commands::FontCommand, functions::Function,
        greeks::Greek, groupings::Grouping, logicals::Logical, operators::Operator, others::Other,
        relations::Relation,
    },
    Span,
};

/// Token identified in the ascii math source.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
pub(crate) struct Token<'src> {
    kind: TokenKind,
    content: &'src str,
    span: Span,
}

impl Token<'_> {
    pub fn with_span(content: &str, kind: TokenKind, span: Span) -> Token<'_> {
        Token {
            content,
            kind,
            span,
        }
    }

    #[allow(dead_code)] // is used in testing
    pub fn span(&self) -> Span {
        self.span
    }

    #[allow(dead_code)] // is used in testing
    pub fn kind(&self) -> TokenKind {
        self.kind
    }

    #[allow(dead_code)] // is used in testing
    pub fn as_str(&self) -> &str {
        self.content
    }
}

/// Kind of token identified in ascii math input.
#[derive(Default, Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub(crate) enum TokenKind {
    /// Number literal.
    Number,

    /// Greek letter, e.g. pi, alpha etc.
    Greek(Greek),

    /// Various arrows, for example |->
    Arrow(Arrow),

    /// Standard functions, eg. sin, cos etc.
    Function(Function),

    /// Standard operators, e.g. +, -, *, |>< etc.
    Operator(Operator),

    /// Symbols that are not letters, and aren't explicitely defined in AsciiMath grammar fall back
    /// to unknown operators.
    UnknownOperator,

    /// Relations in maths, e.g. =, !=, <, <= etc.
    Relation(Relation),

    /// Logic symbols, e.g. and, or, for all etc.
    Logical(Logical),

    /// Grouping brackets in ascii math, e.g. (, {, [, (: etc.
    Grouping(Grouping),

    /// Other (Miscellaneous) symbols of ascii math.
    Other(Other),

    /// Ascii math keywords for accents, e.g. hat, bar, ubrace etc.
    Accent(Accent),

    /// Ascii math font commands such as bb, cc etc.
    FontCommand(FontCommand),

    /// Other Ascii symbol that was not recognized as a token.
    Variable,

    #[default]
    Unimplemented,
}

impl TokenKind {
    pub fn is_grouping_open(&self) -> bool {
        let TokenKind::Grouping(grp) = self else {
            return false;
        };

        !matches!(
            grp,
            Grouping::CloseParen
                | Grouping::CloseBracket
                | Grouping::CloseBrace
                | Grouping::RightAngled
                | Grouping::CloseIgnored
        )
    }
}
