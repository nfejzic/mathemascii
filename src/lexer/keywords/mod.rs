//! Constants containing all the keywords defined in the asciimath grammar.

use crate::scanner::Symbol;

use super::token::TokenKind;

mod macros;

pub mod arrows;
pub mod functions;
pub mod greek;

pub(crate) trait KeywordKind: Into<TokenKind> + Copy {
    fn prefix_of(&self) -> Option<usize>;
}

pub(crate) trait Keyword {
    const MAX_LEN: usize;
    const MIN_LEN: usize;

    type Kind: KeywordKind;

    fn get(key: &str) -> Option<Self::Kind>;
    fn starts_with(symbol: Symbol<'_>) -> bool;
}
