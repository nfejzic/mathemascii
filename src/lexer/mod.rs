//! Tokens, lexer and other helper types and functions for tokenization of asciimath input.

mod next_impl;
mod token;

use crate::scanner::{Scan, Symbol};

use token::{Token, TokenKind};

use self::keywords::{
    accents::Accents, arrows::Arrows, functions::Functions, greek::Greeks, groupings::Groupings,
    logicals::Logicals, operators::Operators, others::Others, relations::Relations, Keyword,
    KeywordKind,
};

mod keywords;

#[derive(Debug, Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub(crate) struct Span {
    start: usize,
    end: usize,
}

/// Iterator that finds and returns tokens in AsciiMath input.
///
/// In cases where a token is prefix of other token, the longer token is given precedence. For
/// example: 'g' is function g, and 'gamma' is greek letter. In order to correctly identify the
/// greek letter, the longer token must have precedence.
#[derive(Default, Debug, Clone, PartialEq, Eq)]
pub(crate) struct TokenIterator<'src> {
    src: Vec<Symbol<'src>>,
    curr: usize,
    span: Span,
}

pub(crate) trait Tokenize {
    fn tokenize(&self) -> TokenIterator;
}

impl<T> Tokenize for T
where
    T: Scan,
{
    fn tokenize(&self) -> TokenIterator {
        TokenIterator {
            src: self.scan(),
            ..Default::default()
        }
    }
}

impl<'src> TokenIterator<'src> {
    fn skip_whitespace(&mut self) {
        loop {
            match self.src.get(self.curr) {
                Some(sym) if sym.is_whitespace() => self.curr += 1,
                _ => break,
            }
        }
    }

    /// Identifies a number in AsciiMath input, like 42, 42.24, .3 etc.
    fn lex_number(&self) -> Option<(Token<'src>, usize)> {
        let mut dot_seen = false;

        let start = self.curr;
        let mut curr = self.curr;

        while let Some(sym) = self.src.get(curr) {
            if !sym.is_digit() && !sym.is_dot() {
                break;
            } else if sym.is_dot() {
                if dot_seen {
                    break;
                }

                dot_seen = true;
            }

            curr += 1;
        }

        if start == curr {
            return None;
        }

        let content = Symbol::as_str(self.src.get(start..curr)?)?;

        if matches!(content, ".") {
            // single dot is not a valid number
            return None;
        }

        let kind = TokenKind::Number;

        let span = Span { start, end: curr };
        Some((Token::with_span(content, kind, span), curr))
    }

    /// Lexes a keyword with a given minimum length (or default for the given keyword if that is
    /// longer)
    fn lex_keyword<K>(&self, min_len: usize) -> Option<(Token<'src>, usize)>
    where
        K: Keyword,
    {
        if let Some(sym) = self.src.get(self.curr) {
            if !<K as Keyword>::starts_with(*sym) {
                // none of the corresponding keywords start with the given symbol, so skip parsing
                return None;
            }
        }

        let min_len = K::MIN_LEN.max(min_len);

        let start = self.curr;
        let mut curr = (self.curr + min_len).saturating_sub(1);
        let mut found_at = start;
        let mut keyword = None;

        while let Some(slice) = self.src.get(start..=curr) {
            if slice.last().map(|s| s.is_whitespace()).unwrap_or(false) {
                // token can't contain a whitespace
                break;
            }

            curr += 1;

            let len = slice.len();

            if len > K::MAX_LEN {
                break;
            }

            let slice_str = Symbol::as_str(slice)?;

            if let Some(kind) = <K as Keyword>::get(slice_str) {
                // longer keywords have precedence, otherwise they would not be possible to lex...
                let span = Span { start, end: curr };
                keyword = Some(Token::with_span(slice_str, kind.into(), span));
                found_at = curr;

                match kind.prefix_of() {
                    Some(word_len) => {
                        if len > word_len {
                            break;
                        }
                    }
                    None => break,
                }
            }
        }

        // keyword lexed up to the `found_at` index. Since we check if longer keywords can be
        // lexed, current index is set beyond this point. To make sure we don't overshoot the
        // index, reset to the position where the keyword was actually lexed.
        keyword.map(|k| (k, found_at))
    }

    /// Identifies a greek letter in AsciiMath input, e.g. alpha, beta, pi, Psi etc.
    fn lex_greek(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Greeks>(min_len)
    }

    /// Identifies an arrow in AsciiMath input, e.g. ->, MapsTo, |-> etc.
    fn lex_arrow(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Arrows>(min_len)
    }

    /// Identifies a function in AsciiMath input, e.g. log, sin, cosh etc.
    fn lex_function(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Functions>(min_len)
    }

    fn lex_operator(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Operators>(min_len)
    }

    fn lex_relation(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Relations>(min_len)
    }

    fn lex_logical(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Logicals>(min_len)
    }

    fn lex_grouping(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Groupings>(min_len)
    }

    fn lex_other(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Others>(min_len)
    }

    fn lex_accent(&self, min_len: usize) -> Option<(Token<'src>, usize)> {
        self.lex_keyword::<Accents>(min_len)
    }
}

impl<'src> Iterator for TokenIterator<'src> {
    type Item = Token<'src>;

    fn next(&mut self) -> Option<Self::Item> {
        self.skip_whitespace();

        match self.src.get(self.curr) {
            Some(_) => {
                next_impl::next_impl!(
                self,
                no_prefix:
                    lex_number;
                prefix:
                    lex_greek,
                    lex_arrow,
                    lex_function,
                    lex_operator,
                    lex_relation,
                    lex_logical,
                    lex_grouping,
                    lex_other,
                    lex_accent
                )
            }
            None => None,
        }
    }
}

#[cfg(test)]
mod tests;
