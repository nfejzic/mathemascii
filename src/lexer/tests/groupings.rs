use super::Snapshot;

use crate::lexer::Tokenize;

super::test_snap!(parens, "()(()))");
super::test_snap!(brackets, "[][[]]][");
super::test_snap!(angled, "langle(::)<<>>");
super::test_snap!(ignored, "( :}{:)");
