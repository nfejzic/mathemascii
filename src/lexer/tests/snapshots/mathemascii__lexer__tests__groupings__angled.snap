---
source: src/lexer/tests/groupings.rs
expression: "Snapshot((input, tokens))"
---
langle(::)<<>>

langle
^^^^^^ -> Grouping(LeftAngled) at: 0 -> 6

      (:
      ^^ -> Grouping(LeftAngled) at: 6 -> 8

        :)
        ^^ -> Grouping(RightAngled) at: 8 -> 10

          <<
          ^^ -> Grouping(LeftAngled) at: 10 -> 12

            >>
            ^^ -> Grouping(RightAngled) at: 12 -> 14


